### ISSUE TITLE
【BUG】When Increase the training epoch, predict_proba would be NaN.

### ISSUE THREAD

**Issue by mylamour** [dnouri/skorch#332]
Recently, i am learning pytorch, also want combine it with sklearn. Fortunately, i found this project. 
it's pretty cool and thank you. But when i try to write demo from tutorial code. this was a problem. when you increase the epoch size, then predict_proba would be NaN. I am not sure what's wrong with it. But you can see it, when epoch size was changed from 100 to 300. this problem was happened.

**test.py**
```python

import torch
import torch.nn.functional as F
import numpy as np

from torch import nn
from skorch import NeuralNetClassifier
from sklearn.datasets import load_digits


digits = load_digits()
X = torch.from_numpy(digits.data.astype(np.float32))
y = torch.from_numpy(digits.target.astype(np.int64))

class Net(nn.Module):
    def __init__(self, num_units=20, nonlin=F.relu):
        super(Net, self).__init__()

        self.dense0 = nn.Linear(64, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(0.5)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 10)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

clf = NeuralNetClassifier(
    Net,
    max_epochs=100,
    lr=0.1
)

clf.fit(X, y)
y_proba = clf.predict_proba(X)
print(y_proba)


```

**max_epochs=100**

```shell
(mldl)  ~/AI/skorch/examples/ [master] python test.py
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        2.2746       0.2857        2.1088  0.0579
      2        2.0081       0.2857        1.9887  0.0526
      3        1.9421       0.3846        1.7229  0.0460
      4        1.6476       0.3846        1.4262  0.0482
      5        1.4475       0.6209        1.0402  0.0513
      6        1.4332       0.6484        1.0375  0.0487
      7        1.3636       0.7802        0.8855  0.0547
      8        1.2248       0.7253        0.8289  0.0502
      9        1.2989       0.6703        1.0695  0.0496
                .....             .........       ........     
                .....             .........       ........    
                .....             .........       ........      

     98        0.6803       0.8984        0.3704  0.0489
     99        0.6764       0.9176        0.3565  0.0458
    100        0.6497       0.8791        0.3491  0.0482
[[7.3202348e-01 2.2077117e-07 1.9366424e-04 ... 3.9416759e-06
  7.0143005e-07 2.5466156e-01]
 [1.3517741e-10 9.9163985e-01 5.0118266e-05 ... 2.2563489e-08
  5.6848279e-03 1.5677013e-08]
 [5.6715899e-10 2.8967354e-01 1.7140906e-01 ... 1.9670969e-07
  5.3394657e-01 1.8424357e-08]
 ...
 [4.2307982e-04 8.9185210e-03 3.9226790e-03 ... 2.3897275e-04
  9.7830141e-01 1.9110983e-06]
 [4.1731731e-03 1.7939850e-03 1.0104966e-03 ... 1.0837655e-03
  4.7035632e-04 9.7788244e-01]
 [2.8054544e-03 1.8803196e-02 2.2230349e-02 ... 3.0519355e-03
  9.1223693e-01 1.3240652e-04]]
(mldl)  ~/AI/skorch/examples/ [master]

```

**max_epochs=300**
```bash
(mldl)  ~/AI/skorch/examples/ [master] python test.py
  epoch    train_loss    valid_acc    valid_loss     dur
-------  ------------  -----------  ------------  ------
      1        2.3342       0.1456        2.2465  0.1157
      2        2.0756       0.4011        1.9379  0.0510
      3        1.8148       0.6676        1.4112  0.0519
      4        1.6672       0.4505        1.4237  0.0511
                .....             .........       ........     
                .....             .........       ........    
                .....             .........       ........      
    117           nan       0.0989           nan  0.0513
    
                .....             .........       ........     
                .....             .........       ........    
                .....             .........       ........      
    
    297           nan       0.0989           nan  0.0725
    298           nan       0.0989           nan  0.0633
    299           nan       0.0989           nan  0.0568
    300           nan       0.0989           nan  0.0559
[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]
(mldl)  ~/AI/skorch/examples/ [master]
```

Eh, I found a new problem, when I try to pass parameter in Model, this was a new problem.

```python

clf = NeuralNetClassifier(
    Net(In=X.shape[1],Out=y.shape[0]),
    max_epochs=100,
    lr=0.1
)

```

![image](https://user-images.githubusercontent.com/12653147/45082101-2f4bb180-b12b-11e8-9c27-caf9a86f9adc.png)

**Comment by thomasjpfan** [comment]
This behavior is known as "Exploding Gradient". You can read more about it here: https://machinelearningmastery.com/exploding-gradients-in-neural-networks/

`skorch.callbacks.GradientNormClipping` can be used to prevent the gradients from getting too big.

**Comment by mylamour** [comment]
Thanks you very much @thomasjpfan , but there also have another question, you can see it in last few lines. Epoch size was same, but obviously,  train_loss change to long since 10 .

**Comment by thomasjpfan** [comment]
Gradient exploding depends on your batch_size and your model. In your case, the model was changed with `In` and `Out` parameters. There is also a random element to it, since the model weights are initialized randomly.

**Comment by mylamour** [comment]
So, there was two differernt model ? even `input` shape and `output` shape was same? I have little confused about it and  which way is better? 

```python
class Net(nn.Module):
    def __init__(self, num_units=20, nonlin=F.relu):
        super(Net, self).__init__()

        self.dense0 = nn.Linear(64, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(0.5)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, 10)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

clf = NeuralNetClassifier(
    Net,
    max_epochs=100,
    lr=0.1
)

```

```python
class Net(nn.Module):
    def __init__(self, In, Out,num_units=20, nonlin=F.relu):
        super(Net, self).__init__()

        self.dense0 = nn.Linear(In, num_units)
        self.nonlin = nonlin
        self.dropout = nn.Dropout(0.5)
        self.dense1 = nn.Linear(num_units, 10)
        self.output = nn.Linear(10, Out)

    def forward(self, X, **kwargs):
        X = self.nonlin(self.dense0(X))
        X = self.dropout(X)
        X = F.relu(self.dense1(X))
        X = F.softmax(self.output(X), dim=-1)
        return X

clf = NeuralNetClassifier(
    Net(In=X.shape[1],Out=y.shape[0]),
    max_epochs=100,
    lr=0.1
)
```

**Comment by benjamin-work** [comment]
These two models look equivalent, so maybe it was just randomness that caused the difference. In general, there is always randomness involved, so make sure to run your experiments repeatedly. Also, try setting random seeds: 

https://pytorch.org/docs/master/torch.html?#torch.manual_seed
https://pytorch.org/docs/stable/cuda.html#torch.cuda.manual_seed

Moreover, in this kind of situation, lower learning rates often help.

**Comment by mylamour** [comment]
Thanks very much, @thomasjpfan  @benjamin-work .
